Sources/OpenLogin/WebAuth.swift

```diff
@@ -47,75 +91,115 @@ public class WebAuth: NSObject {
 
      - parameter callback: Callback called with the result of the WebAuth flow.
      */
-    public func start(_ callback: @escaping (Result<State>) -> Void) {
-        guard
-            let bundleId = Bundle.main.bundleIdentifier,
-            let redirectURL = URL(string: "\(bundleId)://openlogin")
-        else { return callback(.failure(WebAuthError.noBundleIdentifierFound)) }
+    public func login(_ loginParams: OLLoginParams, _ callback: @escaping (Result<OpenLoginState>) -> Void) {
+        DispatchQueue.main.async { [self] in
+            guard
+                let bundleId = Bundle.main.bundleIdentifier,
+                let redirectURL = URL(string: "\(bundleId)://openlogin")
+            else { return callback(.failure(WebAuthError.noBundleIdentifierFound)) }
+            
+            guard
+                let url = try? WebAuth.generateAuthSessionURL(redirectURL: redirectURL, initParams: initParams, loginParams: loginParams)
+            else {
+                return callback(.failure(WebAuthError.unknownError))
+            }
+            
+            let authSession = ASWebAuthenticationSession(
+                url: url, callbackURLScheme: redirectURL.scheme) { callbackURL, authError in
+                guard
+                    authError == nil,
+                    let callbackURL = callbackURL,
+                    let callbackState = try? WebAuth.decodeStateFromCallbackURL(callbackURL)
+                else {
+                    let authError = authError ?? WebAuthError.unknownError
+                    if case ASWebAuthenticationSessionError.canceledLogin = authError {
+                        return callback(.failure(WebAuthError.userCancelled))
+                    } else {
+                        return callback(.failure(authError))
+                    }
+                }
+                callback(.success(callbackState))
+            }
+            
+            if #available(iOS 13.0, *) {
+                authSession.presentationContextProvider = self
+            }
+            
+            if !authSession.start() {
+                callback(.failure(WebAuthError.unknownError))
+            }
+        }
+    }
+    
+    static func generateAuthSessionURL(redirectURL: URL, initParams: OLInitParams, loginParams: OLLoginParams) throws -> URL {
+        
+        var sdkParams: Dictionary<String, Any> = [:]
+        
```

Comment:
do you not have a gson type library (which does serialization to json directly?)

Reply:
no gson, but should have somehing similar, i just copied the existing code, let me refactor

Reply:
turns out impossible since there is a Dict, Swift's gson won't work with that.